mount/mount_linux.go

```diff
@@ -292,18 +298,74 @@ func (mounter *SafeFormatAndMount) checkAndRepairXfsFilesystem(source string) er
 			return nil
 		} else {
 			klog.Warningf("Filesystem corruption was detected for %s, running xfs_repair to repair", source)
-			out, err := mounter.Exec.Command("xfs_repair", args...).CombinedOutput()
-			if err != nil {
-				return NewMountError(HasFilesystemErrors, "'xfs_repair' found errors on device %s but could not correct them: %s\n", source, out)
-			} else {
-				klog.Infof("Device %s has errors which were corrected by xfs_repair.", source)
-				return nil
+			var uncorrectedErr error
+
+			// If xfs_repair fails to repair the file system successfully, try giving the same xfs_repair
+			// command twice more; xfs_repair may be able to make more repairs on successive runs.
+			for i := 0; i < xfsRepairRetries; i++ {
+				out, err := mounter.Exec.Command("xfs_repair", args...).CombinedOutput()
+				if err == nil {
+					klog.Infof("Device %s has errors which were corrected by xfs_repair.", source)
+					return nil
+				}
+				e, isExitError := err.(utilexec.ExitError)
+				if !isExitError {
+					return NewMountError(HasFilesystemErrors, "failed to run 'xfs_repair' on device %s: %v\n", source, err)
+
+				}
+				switch e.ExitStatus() {
+				case xfsRepairErrorsUncorrected:
+					uncorrectedErr = NewMountError(HasFilesystemErrors, "'xfs_repair' found errors on device %s but could not correct them: %s\n", source, string(out))
+					// If the exit status is 2, do not retry, replay the dirty logs instead.
+				case xfsRepairErrorsDirtyLogs:
+					return mounter.replayXfsDirtyLogs(source, target)
+				default:
+					return NewMountError(HasFilesystemErrors, "'xfs_repair' found errors on device %s but could not correct them: %s\n", source, string(out))
+				}
+			}
+			if uncorrectedErr != nil {
+				return uncorrectedErr
 			}
 		}
 	}
 	return nil
 }
 
+// replayXfsDirtyLogs tries to replay dirty logs by by mounting and
+// immediately unmounting the filesystem on the same class of machine
+// that crashed.
+func (mounter *SafeFormatAndMount) replayXfsDirtyLogs(source, target string) error {
+	klog.V(4).Infof("Attempting to replay the dirty logs on device %s", source)
+	msg := fmt.Sprintf("failed to replay dirty logs on device %s", source)
+
+	// make sure unmount is always called after mount.
+	err := func() (returnErr error) {
+		defer func() {
+			klog.V(4).Infof("Unmounting %s", target)
+			if err := mounter.Interface.Unmount(target); err != nil {
+				returnErr = NewMountError(HasFilesystemErrors, "%s: %v\n", msg, err)
+			}
+		}()
+		klog.V(4).Infof("Attempting to mount disk %s at %s", source, target)
+		if err := mounter.Interface.Mount(source, target, "", []string{"defaults"}); err != nil {
```

Comment:
From man-page:

```
In this situation, the log can be replayed by mounting and
       immediately unmounting the filesystem on the same class of machine
       that crashed.  Please make sure that the machine's hardware is
       reliable before replaying to avoid compounding the problems.
```

Are we being too aggressive in automatically fixing errors here? Can this make problems worse somehow? Should this be configurable? If we merge this PR as it is - it will become the new default even for in-tree Kubernetes drivers, so we have to be careful.

Reply:
@gnufied do you mean that the node might be bad or not the same class, so there is potential risk?